### PR TITLE
[feat] Add support for the Apptainer container platform

### LINKED ISSUE #2602: Support apptainer container platform
As far as I understand from @teojgo , it should be very easy to add as we already support Singularity.

### PR DESCRIPTION
**Pull request by teojgo**
Fixes #2602

### CODE REVIEW

**Review by rsarm — APPROVED** [review]

File: reframe/core/containers.py [hunk 1]
```diff
@@ -259,10 +260,19 @@ def launch_command(self, stagedir):
...
+            return (f'{self._launch_command} exec {" ".join(run_opts)} '
                     f'{self.image} {self.command}')
 
-        return f'singularity run {" ".join(run_opts)} {self.image}'
+        return f'{self._launch_command} run {" ".join(run_opts)} {self.image}'
+
+
+class Apptainer(Singularity):
+    '''Container platform backend for running containers with `Apptainer
+    <https://apptainer.org/>`__.'''
```
Comment (vkarak) [inline reframe/core/containers.py:271]: Add an annotation here that this was added in version 4.0.0.

File: unittests/test_containers.py [hunk 2]
```diff
@@ -101,7 +102,7 @@ def expected_cmd_mount_points(container_variant):
                 '--mount=type=bind,source="/path/two",destination="/two" '
                 '--mount=type=bind,source="/foo",destination="/rfm_workdir" '
                 'load/library/image:tag cmd')
-    elif container_variant in {'Singularity', 'Singularity+nopull'}:
+    elif container_variant in {'Singularity'}:
```
Comment (vkarak) [inline unittests/test_containers.py:105]:
```suggestion
    elif container_variant == 'Singularity':
```

File: unittests/test_containers.py [hunk 3]
```diff
@@ -189,6 +203,15 @@ def expected_cmd_run_opts(container_variant):
     elif container_variant == 'Singularity+nocommand':
         return ('singularity run -B"/path/one:/one" -B"/foo:/rfm_workdir" '
                 '--foo --bar image:tag')
+    elif container_variant in {'Apptainer'}:
```
Comment (vkarak) [inline unittests/test_containers.py:206]:
```suggestion
    elif container_variant == 'Apptainer':
```

**Review by vkarak — CHANGES_REQUESTED** [review]

File: reframe/core/containers.py (hunk 1 above)
Comment (vkarak) [reply reframe/core/containers.py:271]: You need to list it also here: https://reframe-hpc.readthedocs.io/en/stable/config_reference.html#systems-.partitions-.container_platforms-.type

File: unittests/test_containers.py [hunk 4]
```diff
@@ -13,7 +13,8 @@
     'Sarus', 'Sarus+nocommand', 'Sarus+nopull', 'Sarus+mpi', 'Sarus+load',
     'Shifter', 'Shifter+nocommand', 'Shifter+mpi', 'Shifter+nopull',
     'Shifter+load',
-    'Singularity', 'Singularity+nocommand', 'Singularity+cuda'
+    'Singularity', 'Singularity+nocommand', 'Singularity+cuda',
+    'Apptainer', 'Apptainer+nocommand', 'Apptainer+cuda'
```
Comment (ekouts) [inline unittests/test_containers.py:17]: It looks like these params are in alphabetical order, but it's fine to leave like that

**Review by ekouts — APPROVED** [review]

File: unittests/test_containers.py (hunk 4 above)
Comment (teojgo) [reply unittests/test_containers.py:17]: I was thinking to change them but in the end I left it like that to be side by side with the Singularity one

**Review by vkarak — APPROVED** [review]

**Pull request merged**